src/main/java/io/netty/buffer/api/BufHolder.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+import java.lang.invoke.VarHandle;
+import java.util.Objects;
+
+import static io.netty.buffer.api.Statics.findVarHandle;
+import static java.lang.invoke.MethodHandles.lookup;
+
+/**
+ * The {@link BufHolder} is an abstract class that simplifies the implementation of objects that themselves contain
+ * a {@link Buf} instance.
+ * <p>
+ * The {@link BufHolder} can only hold on to a single buffer, so objects and classes that need to hold on to multiple
+ * buffers will have to do their implementation from scratch, though they can use the code of the {@link BufHolder} as
+ * inspiration.
+ * <p>
+ * If you just want an object that is a reference to a buffer, then the {@link BufRef} can be used for that purpose.
+ * If you have an advanced use case where you wish to implement {@link Rc}, and tightly control lifetimes, then
+ * {@link RcSupport} can be of help.
+ *
+ * @param <T> The concrete {@link BufHolder} type.
+ */
+public abstract class BufHolder<T extends BufHolder<T>> implements Rc<T> {
+    private static final VarHandle BUF = findVarHandle(lookup(), BufHolder.class, "buf", Buf.class);
+    private Buf buf;
```

Comment:
Does this need to be `volatile` ? We use `getAndSet(...)` below so I would suspect it must be volatile.

Reply:
I've added methods to expose both plain and volatile access modes for this field.